eli5/keras.py

```diff
@@ -0,0 +1,465 @@
+    else:
+        raise ValueError('Could not find a suitable target layer automatically.')
+
+
+def is_suitable_activation_layer(estimator, i):
+    # type: (Model, int) -> bool
+    """
+    Check whether
+    the layer at index ``i`` matches what is required 
+    by ``estimator``.
+    
+    Parameters
+    ----------
+    estimator : keras.models.Model
+        Model from which to retrieve the layer.
+        See :func:`explain_prediction_keras`.
+
+    i : int
+        Index into the ``estimator``'s layers list.
+
+    Returns
+    -------
+    is_suitable : boolean
+        whether the layer matches what is needed
+
+    Notes
+    -----
+
+    Matching Criteria
+        * Rank of the layer's output tensor.
+    """
+    # TODO: experiment with this, using many models and images, to find what works best
+    # Some ideas: 
+    # check layer type, i.e.: isinstance(l, keras.layers.Conv2D)
+    # check layer name
+    l = estimator.get_layer(index=i)
+    # a check that asks 'can we resize this activation layer over the image?'
+    rank = len(l.output_shape)
+    required_rank = len(estimator.input_shape)
+    return rank == required_rank
+
+
+def grad_cam(estimator, doc, targets, activation_layer):
+    # type: (Model, np.ndarray, Optional[list], Layer) -> Tuple[np.ndarray, int, float]
+    """
+    Generate a heatmap using Gradient-weighted Class Activation Mapping 
+    (Grad-CAM) (https://arxiv.org/pdf/1610.02391.pdf).
+    
+    See :func:`explain_prediction_keras` for more information about the 
+    ``estimator``, ``doc`` and ``targets`` parameters.
+
+    Parameters
+    ----------
+    estimator : keras.models.Model
+        Model to Grad-CAM on.
+
+    doc : numpy.ndarray
+        Input image to ``estimator``.
+
+    targets : list or None
+        Prediction to focus on that can be made by ``estimator``.
+
+    activation_layer : keras.layers.Layer
+        Hidden layer in ``estimator`` to differentiate with respect to.
+
+    
+    Returns
+    -------
+    (heatmap, predicted_idx, score) : (numpy.ndarray, int, float)
+        A Grad-CAM localization map,
+        the predicted class ID, and
+        the score (i.e. probability) for that class.
+
+
+    Notes
+    -----
+    We currently make two assumptions in this implementation
+        * We are dealing with images as our input to ``estimator``.
+        * We are doing a classification. Our ``estimator``'s output is a class scores vector.
+
+    Credits
+        * Jacob Gildenblat for "https://github.com/jacobgil/keras-grad-cam".
+        * Author of "https://github.com/PowerOfCreation/keras-grad-cam" for fixes to Jacob's implementation.
+        * Kotikalapudi, Raghavendra and contributors for "https://github.com/raghakot/keras-vis".
+    """
+    # Get required terms
+    weights, activations, grads, predicted_idx, score = grad_cam_backend(estimator, doc, targets, activation_layer)
+
+    # TODO: might want to replace np operations with keras backend operations
```

Comment:
I think it's fine to have them as numpy - when we add pytorch support in the future, this code would likely be re-used.